### DISCUSSION TITLE
How to register COM DLL in v5?

### DISCUSSION THREAD

**Discussion by siddhsql** [wixtoolset/issues#8820] [Questions]
Is there any page that describes how to register a COM DLL in WIX v5?

**Comment by wmanning** [comment]
This maybe: https://stackoverflow.com/a/364544/623561
I don't think that has changed since Wix 3.x.

**Comment by siddhsql** [comment]
I read that. the problem is heat.exe (even v5) cannot handle 64 bit DLLs.
also on the SO answer they strongly advise against self-registration yet on
firegiant website they seem to advocate for it:
https://www.firegiant.com/docs/fg-wix/harvesting/selfreg/
what is the difference if any between using this:

<Component Directory="INSTALLFOLDER"> <fg:HarvestFile Source=
"!(bindpath.Foo)\path\to\foo.dll" SelfReg="yes" /> </Component>

vs. creating a custom action that calls regsvr32.exe to register the DLL?

**Reply by robmen** [reply]
heat was never designed to harvest 64-bit selfreg, just like it doesn't support out-of-proc harvesting. The heat code is ancient, annoying, and uncared for. For those reasons, it will be [removed from WiX in a future release](https://github.com/wixtoolset/issues/issues/6838).

At FireGiant, we had customers that needed those harvesting features, so we created [Advanced Harvesting](https://www.firegiant.com/docs/heatwave-build-tools/harvesting/) to tackle the challenges of 64-bit selfreg and out-of-proc harvesting plus many, many more.

**Comment by siddhsql** [comment]
so what's the solution to the OP then? and is wix toolset an outdated
technology? should i be using something else to create an installer for a
windows desktop app? i am creating an installer for the first time and
would appreciate some guidance.

**Reply by robmen** [reply]
There are many options. 

1. You type the necessary `RegistryValue` elements to create the registry keys or use the syntactic sugar elements like `Class` and `Interface` and `TypeLib` elements (which can create the same registry keys in the end but with less typing). To do this, you will need to understand the COM registry keys, and MSDN can teach you those.

2. If you don't want to learn about COM registration or you don't want to type all the registry keys, you can buy the Advanced Harvesting tools from FireGiant to do the work for you.

There are probably other options I'm not thinking of.

In the end, I don't know how you get from "I want the WiX Toolset to do everything for me for free no matter how complex it is to make that happen" to "is wix toolset an outdated technology?"

My guidance is to either a) pay someone to help you or b) take the time to read about COM (on MSDN), MSI (on MSDN) and WiX (on wixtoolset.org and firegiant.com) so you have a basic understanding of the tasks ahead of you.

**Comment by chrpai** [comment]
3. SelfReg="1"

4. Stop using COM. (My favorite.)
